tree/ntuple/v7/inc/ROOT/RCluster.hxx

```diff
@@ -0,0 +1,188 @@
+#include <vector>
+
+#ifndef ROOT7_RCluster
+#define ROOT7_RCluster
+
+namespace ROOT {
+namespace Experimental {
+namespace Detail {
+
+
+// clang-format off
+/**
+\class ROOT::Experimental::ROnDiskPage
+\ingroup NTuple
+\brief A page as being stored on disk, that is packed and compressed
+
+Used by the cluster pool to cache pages from the physical storage. Such pages generally need to be
+uncompressed and unpacked before they can be used by RNTuple upper layers.
+*/
+// clang-format on
+class ROnDiskPage {
+private:
+   /// The memory location of the bytes
+   const void *fAddress = nullptr;
+   /// The compressed and packed size of the page
+   std::size_t fSize = 0;
+
+public:
+   /// On disk pages within a page source are identified by the column and page number. The key is used for
+   /// associative collections of on disk pages.
+   struct Key {
+      DescriptorId_t fColumnId;
+      NTupleSize_t fPageNo;
+      Key(DescriptorId_t columnId, NTupleSize_t pageNo) : fColumnId(columnId), fPageNo(pageNo) {}
+      bool operator ==(const Key &other) const {
+         return fColumnId == other.fColumnId && fPageNo == other.fPageNo;
+      }
+   };
+
+   ROnDiskPage() = default;
+   ROnDiskPage(void *address, std::size_t size) : fAddress(address), fSize(size) {}
+
+   const void *GetAddress() const { return fAddress; }
+   std::size_t GetSize() const { return fSize; }
+
+   bool IsNull() const { return fAddress == nullptr; }
+};
+
+} // namespace Detail
+} // namespace Experimental
+} // namespace ROOT
+
+// For hash maps ROnDiskPage::Key --> ROnDiskPage
+namespace std
+{
+   template <>
+   struct hash<ROOT::Experimental::Detail::ROnDiskPage::Key>
+   {
+      // TODO(jblomer): quick and dirty hash, likely very sub-optimal, to be revised later.
+      size_t operator()(const ROOT::Experimental::Detail::ROnDiskPage::Key &key) const
+      {
+         return ((std::hash<ROOT::Experimental::DescriptorId_t>()(key.fColumnId) ^
+                 (hash<ROOT::Experimental::NTupleSize_t>()(key.fPageNo) << 1)) >> 1);
+      }
+   };
+}
+
+
+namespace ROOT {
+namespace Experimental {
+namespace Detail {
+
+// clang-format off
+/**
+\class ROOT::Experimental::Detail::ROnDiskPageMap
+\ingroup NTuple
+\brief A memory region that contains packed and compressed pages
+
+Derived classes implement how the on disk pages are stored in memory, e.g. mmap'd or in a special area.
+*/
+// clang-format on
+class ROnDiskPageMap {
+   friend class RCluster;
+
+protected:
+   /// The memory region containing the on-disk pages. Ownership of the memory region is passed to the page map.
+   /// Therefore, the region needs to be allocated in a way that fits the derived class and its destructor.
+   void *fMemory;
+   std::unordered_map<ROnDiskPage::Key, ROnDiskPage> fOnDiskPages;
+
+public:
+   explicit ROnDiskPageMap(void *memory) : fMemory(memory) {}
+   ROnDiskPageMap(const ROnDiskPageMap &other) = delete;
+   ROnDiskPageMap(ROnDiskPageMap &&other);
```

Comment:
As this is a perf optimization, inlining it will be crucial.

Reply:
I added a clarifying comment.  It's not primarily a performance optimization but the default move constructor and assignment do not correctly manage the `fMemory` memory area.

Reply:
Hm, then perhaps I don't understand.  The reason for this custom move constructor is not performance but correctness.  I can of course inline but this is not a performance critical spot and perhaps would unnecessarily grow the header.

Reply:
Oh, probably it doesn't even matter, as `fMemory` should any go as a unique char array to the heap page map.